cpp/src/parquet/encoding.cc

```diff
@@ -2236,6 +2326,78 @@ class DeltaByteArrayDecoder : public DecoderImpl,
   ByteArray last_value_;
 };
 
+// ----------------------------------------------------------------------
+// BYTE_STREAM_SPLIT
+
+template <typename DType>
+class ByteStreamSplitDecoder : public DecoderImpl, virtual public TypedDecoder<DType> {
+ public:
+  using T = typename DType::c_type;
+  explicit ByteStreamSplitDecoder(const ColumnDescriptor* descr);
+
+  int Decode(T* buffer, int max_values) override;
+
+  int DecodeArrow(int num_values, int null_count, const uint8_t* valid_bits,
+                  int64_t valid_bits_offset,
+                  typename EncodingTraits<DType>::Accumulator* builder) override;
+
+  int DecodeArrow(int num_values, int null_count, const uint8_t* valid_bits,
+                  int64_t valid_bits_offset,
+                  typename EncodingTraits<DType>::DictAccumulator* builder) override;
+
+  void SetData(int num_values, const uint8_t* data, int len) override;
+
+ private:
+  int num_values_{0U};
+  int num_values_decoded_{0U};
+};
+
+template <typename DType>
+ByteStreamSplitDecoder<DType>::ByteStreamSplitDecoder(const ColumnDescriptor* descr)
+    : DecoderImpl(descr, Encoding::BYTE_STREAM_SPLIT) {}
+
+template <typename DType>
+void ByteStreamSplitDecoder<DType>::SetData(int num_values, const uint8_t* data,
+                                            int len) {
+  DecoderImpl::SetData(num_values, data, len);
+  num_values_ = num_values;
+  num_values_decoded_ = 0;
+}
+
+template <typename DType>
+int ByteStreamSplitDecoder<DType>::Decode(T* buffer, int max_values) {
+  constexpr size_t num_streams = sizeof(T);
+  using UnsignedType = typename UnsignedTypeWithWidth<sizeof(T)>::type;
+  const int values_to_decode = std::min(num_values_ - num_values_decoded_, max_values);
+  for (int i = 0; i < values_to_decode; ++i) {
+    UnsignedType reconstructed_value_as_uint{0U};
+    for (size_t b = 0; b < num_streams; ++b) {
+      const size_t byte_index = b * num_values_ + num_values_decoded_ + i;
+      const UnsignedType byte_value = static_cast<UnsignedType>(data_[byte_index]);
+      reconstructed_value_as_uint |= (byte_value << (8U * b));
+    }
+    const T reconstructed_value_as_fp =
+        *reinterpret_cast<const T*>(&reconstructed_value_as_uint);
+    buffer[i] = reconstructed_value_as_fp;
+  }
+  num_values_decoded_ += values_to_decode;
```

Comment:
The way it's done in other decoder is `num_values_ -= values_to_decode`. Not sure why you're doing it differently here?

Reply:
It seems like you didn't answer this question. Unless there's a reason we probably shouldn't deviate from how other decoders are implemented.

Reply:
Indeed looks incorrect

Reply:
I missed the question, sorry.

I am using num_values_decoded_ as an offset within each stream of bytes.
Os if I have processed 2 elements in a stream of floats and num_values=5, then the encoded data buffer would be:
XXAAAXXBBBXXCCCXXDDD
where X means used and A,B,C,D are undecoded data for all 4 streams.
So, num_values_decoded_ is 2 here and tells me where I should be looking at.

Why is it necessary that num_values_ is decremented instead of keeping it const and advancing the number of decoded elements?

Reply:
I see. OK, I think this is OK. We might consider changing all the decoders to use this pattern at some point but we could do that later

Reply:
Ah, but I think I have to get it in the same style as the already existing implementation. I will rewrite my patch.